### PR TITLE
Parallelize functions with `rayon`

### PR DESCRIPTION
**Pull request by Milo123459**
This adds a new dependency, `rayon`, that allows us to parallelize iteration over Vecs.

### CODE REVIEW

**Comment by nebulatgs** [comment]
Does this provide a meaningful performance improvement?

**Comment by Milo123459** [comment]
Yes, where things are parallel is often where there is a lot of data going in.

File: src/nixpacks/nix/pkg.rs [hunk 1]
```diff
@@ -22,7 +23,7 @@ impl Pkg {
         match &self.overrides {
             Some(overrides) => {
                 let override_string = overrides
-                    .iter()
+                    .par_iter()
                     .map(|(name, value)| format!("{} = {};", name, value))
```
Comment (coffee-cup) [inline src/nixpacks/nix/pkg.rs:27]: I don't think we should be using rayon for small cases like this. `overrides` is at most going to be length 3 and all the map is doing is formating a string. I think the overhead of rayon will actually be slower.

**Review by coffee-cup — COMMENTED** [review]
Can you notice any significant speed improvement from using rayon? I am not opposed to adding it but feels like most of the cases it is being used in this PR will have at most a vector of length 1-10. Not sure if it is worth adding the extra dependency for that.

File: src/nixpacks/nix/pkg.rs (hunk 1 above)
Comment (Milo123459) [reply src/nixpacks/nix/pkg.rs:27]: Ah, thought overrides was going to be longer then that.

**Comment by nebulatgs** [comment]
> Not sure if it is worth adding the extra dependency for that.

It is most definitely not worth it. Nixpacks is already faster than Docker can read a Dockerfile and produce a build context, so there is no reason to add another dep to (maybe) make it a few milliseconds faster.

**Comment by Milo123459** [comment]
I cannot test locally, however, the largest effect will happen when a large number of environment variables are present. Parallel iteration will help massively.

**Comment by Milo123459** [comment]
Rayon is already in the Cargo sourcetree, this isn't actually adding another dep.

**Comment by coffee-cup** [comment]
> I cannot test locally, however, the largest effect will happen when a large number of environment variables are present. Parallel iteration will help massively.

I think we should test this before merging it in. All that is happening it those iter maps is formatting a string. The speed to map over and format 100 variables is negligible compared to building the image with Docker.



> Rayon is already in the Cargo sourcetree, this isn't actually adding another dep.

Yes but it doesn't appear in the final binary now.

**Pull request closed without merging**